src/main/java/com/minecolonies/coremod/colony/crafting/CustomRecipeManager.java

```diff
@@ -124,12 +126,70 @@ public void reset()
 
     /**
      * Get the custom recipes for an item, or null if no matching recipe exists.
+     * @param item     An individual item to search for recipes.
+     * @return  A list of custom recipes with that output.
      */
+    @Nullable
     public List<CustomRecipe> getRecipeByOutput(final Item item)
     {
         return recipeOutputMap.get(item);
     }
 
+    /**
+     * Gets the custom recipes for an ItemStack, including comparing count and tags, or null if no matching recipe exists.
+     * @param itemStack An ItemStack to search for recipes.
+     * @return  A list of custom recipes with that output.
+     */
+    @Nullable
+    public List<CustomRecipe> getRecipeByOutput(final ItemStack itemStack)
+    {
+        List<CustomRecipe> returnList = new ArrayList<>();
+        for(CustomRecipe recipe : recipeOutputMap.get(itemStack.getItem()))
+        {
+            // ItemStacks don't override equals, so have to use the static methods.
+            if(ItemStack.areItemStacksEqual(recipe.getPrimaryOutput(), itemStack))
+            {
+                returnList.add(recipe);
+            }
+            for(ItemStack output : recipe.getAltOutputs())
+            {
+                if(ItemStack.areItemStacksEqual(output, itemStack))
+                {
+                    returnList.add(recipe);
+                }
+            }
+        }
+        if(returnList.size() > 0)
+        {
+            return returnList;
+        }
+        return null;
+    }
+
+    /**
+     * Gets the custom recipes for an ItemStorage, optionally including comparing count, damage, and NBT, or null if no matching recipe exists.
+     * @param itemStorage An ItemStorage to search for recipes.
+     * @return  A list of custom recipes with that output.
+     */
+    @Nullable
+    public List<CustomRecipe> getRecipeByOutput(final ItemStorage itemStorage)
+    {
+        List<CustomRecipe> returnList = new ArrayList<>();
+        for(CustomRecipe recipe : recipeOutputMap.get(itemStorage.getItem()))
+        {
+            // ItemStorage#equals does the actual comparison work for us, here.
+            if(recipe.getPrimaryOutput().equals(itemStorage) || recipe.getAltOutputs().contains(itemStorage))
+            {
+                returnList.add(recipe);
+            }
+        }
+        if(returnList.size() > 0)
+        {
+            return returnList;
+        }
+        return null;
```

Comment:
Why not just return an empty list here? Why force both a null and a isempty check instead of just the isempty check?